File: crime-log-backend/src/main/java/io/github/aj316/crimelog/backend/model/CrimeReport.java
package io.github.aj316.crimelog.backend.model;

import io.github.aj316.crimelog.backend.model.types.CrimeReportStatus;
import jakarta.persistence.*;
import lombok.Getter;
import lombok.Setter;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.time.LocalDateTime;

@Getter
@Setter
@Entity
@Table(name = "crime_reports")
public class CrimeReport {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long reportId;

    @Column(nullable = false, length = 200)
    private String title;

    @Column(nullable = false, length = 4000)
    private String description;

    @Column(nullable = false, length = 255)
    private String location;

    private LocalDateTime incidentDateTime;

    @Column(nullable = false)
    private Long reporterUserId;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private CrimeReportStatus status = CrimeReportStatus.SENT_TO_POLICE;

    @Column(nullable = false)
    private Boolean publicBroadcasted = false;

    private LocalDateTime acknowledgedAt;
    private Long acknowledgedByUserId;

    private LocalDateTime publicBroadcastedAt;
    private Long publicBroadcastedByUserId;

    @CreationTimestamp
    private LocalDateTime createdAt;

    @UpdateTimestamp
    private LocalDateTime updatedAt;
}

File: crime-log-frontend/src/api/dtos/crime-report.ts
import type {AlertSeverity, CrimeReportStatus, Role} from "../types.ts";

export interface CrimeReportTimelineDto {
    status: CrimeReportStatus;
    note: string | null;
    changedByUserId: number;
    changedByName: string;
    changedByRole: Role;
    createdAt: string;
}

export interface CrimeReportSummaryDto {
    reportId: number;
    title: string;
    location: string;
    status: CrimeReportStatus;
    reporterUserId: number;
    reporterName: string;
    incidentDateTime: string | null;
    createdAt: string;
    updatedAt: string;
    acknowledgedAt: string | null;
    publicBroadcasted: boolean;
}

export interface CrimeReportDetailDto {
    reportId: number;
    title: string;
    description: string;
    location: string;
    status: CrimeReportStatus;
    reporterUserId: number;
    reporterName: string;
    incidentDateTime: string | null;
    createdAt: string;
    updatedAt: string;
    acknowledgedAt: string | null;
    acknowledgedByName: string | null;
    publicBroadcasted: boolean;
    publicBroadcastedAt: string | null;
    publicBroadcastedByName: string | null;
    timeline: CrimeReportTimelineDto[];
}

export interface CreateCrimeReportRequest {
    title: string;
    description: string;
    location: string;
    incidentDateTime: string | null;
}

export interface UpdateCrimeReportStatusRequest {
    status: CrimeReportStatus;
    note?: string | null;
}

export interface BroadcastCrimeReportRequest {
    message?: string | null;
    severity?: AlertSeverity | null;
}
